mlflow/sklearn/__init__.py

```diff
@@ -884,12 +1206,76 @@ def patched_fit(original, self, *args, **kwargs):
                           for autologging (e.g., specify "fit" in order to indicate that
                           `sklearn.linear_model.LogisticRegression.fit()` is being patched)
         """
+        status = _get_autolog_training_status()
         with _SklearnTrainingSession(clazz=self.__class__, allow_children=False) as t:
             if t.should_log():
-                return fit_mlflow(original, self, *args, **kwargs)
+                with status.disable_log_post_training_metrics():
+                    result = fit_mlflow(original, self, *args, **kwargs)
+                status.register_model(self, mlflow.active_run().info.run_id)
+                return result
             else:
                 return original(self, *args, **kwargs)
 
+    def patched_predict(original, self, *args, **kwargs):
+        status = _get_autolog_training_status()
+        if status.should_log_post_training_metrics() and status.get_run_id_for_model(self):
+            predict_result = original(self, *args, **kwargs)
+            eval_dataset = status.get_method_first_arg_value(original, args, kwargs)
+            eval_dataset_name = status.register_eval_dataset(self, eval_dataset)
+            status.register_prediction_result(
+                self, eval_dataset_name, predict_result)
+            return predict_result
+        else:
+            return original(self, *args, **kwargs)
+
+    def patched_metric_api(original, *args, **kwargs):
+        status = _get_autolog_training_status()
+        if status.should_log_post_training_metrics():
+            # one metric api may call another metric api,
+            # to avoid this, call disable_log_post_training_metrics to avoid nested patch
+            with status.disable_log_post_training_metrics():
+                metric = original(*args, **kwargs)
+
+            if status.is_metrics_value_loggable(metric):
+                call_command = status.gen_metric_call_command(None, original, *args, **kwargs)
+
+                is_register_ok, run_id, metric_key = \
+                    status.register_metric_api_call(metric_name, call_command, args, kwargs)
+                if is_register_ok:
+                    status.log_eval_metric(run_id, metric_key, metric)
+
+            return metric
+        else:
+            return original(*args, **kwargs)
+
+    # we need patch model.score method because:
+    #  some model.score() implementation won't call metric APIs in `sklearn.metrics`
+    #  e.g.
+    #  https://github.com/scikit-learn/scikit-learn/blob/82df48934eba1df9a1ed3be98aaace8eada59e6e/sklearn/covariance/_empirical_covariance.py#L220
+    def patched_model_score(original, self, *args, **kwargs):
+        status = _get_autolog_training_status()
+        if status.should_log_post_training_metrics() and status.get_run_id_for_model(self):
+            with status.disable_log_post_training_metrics():
```

Comment:
Can we add an inline comment explaining why we're disabling metric logging here?